library/std/src/f64/soft.rs

```diff
@@ -0,0 +1,352 @@
+//! Software floating point functions for `f64`.
+
+#[cfg(test)]
+mod tests;
+
+/// The floating point type.
+type FP = f64;
+
+/// Type of the exponent.
+type Exponent = i32;
+
+/// Type of `FP::to_bits`.
+type Bits = u64;
+
+/// Half as many bits as `Bits`.
+type HalfBits = u32;
+
+/// Twice as many bits as `Bits`.
+type DoubleBits = u128;
+
+/// Number of bits in the significand.
+const SIGNIFICAND_BITS: u32 = FP::MANTISSA_DIGITS - 1;
+
+/// Number of bits in the exponent.
+const EXPONENT_BITS: u32 = Bits::BITS - 1 - SIGNIFICAND_BITS;
+
+/// Encoded exponent = EXPONENT_BIAS + actual exponent.
+const EXPONENT_BIAS: Exponent = 2 - FP::MIN_EXP;
+
+/// Represents an `FP` number.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+struct Representation {
+    /// Sign.
+    sign: Sign,
+    /// Absolute value.
+    abs: Positive,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Sign {
+    Positive,
+    Negative,
+}
+
+/// Represents a positive number.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Positive {
+    Zero,
+    Finite(PositiveFinite),
+    Infinity,
+    NaN,
+}
+
+/// Represents a non-zero, positive finite number.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+struct PositiveFinite {
+    /// `number = 2^exp * mantissa`.
+    exp: Exponent,
+    /// `2^SIGNIFICAND_BITS <= mantissa < 2^(SIGNIFICAND_BITS + 1)`
+    mantissa: Bits,
+}
```

Comment:
This PR winds up repeating these types a number of times... can it just be made generic? If we need to have separate `impl` blocks for each concrete combination of params, so be it, but we could at least have a unified declaration in an adjacent `pub(crate) mod float;`